Star Wars Unlimited plugin

ThisJustZack opened the pull request:
This PR would allow the Star Wars Unlimited TCG to have plugin support for SCM.

ThisJustZack committed:
Initial commit

MVP for Star Wars Unlimited

Exports only front images of cards for the JSON, Melee, and Picklist formats

ThisJustZack committed:
Initial attempt to add back art

Currently fails completely on JSON for images with this inclusion, so work is needed

ThisJustZack committed:
Modify back image logic and fix possible syntax

Back images now ping swudb before retrieving the image for when the it is unknown if the image exists

Break apart URLs to prevent possible syntax errors

ThisJustZack committed:
Correct back image directory, auto align, Fix json retrieval

Change the image directory for back images to the double-sided directory

Automatically align the images by rotating 90 degrees

Fix JSON issue with leader and base images not being retrieved due to removing the logic for those edge cases

ThisJustZack committed:
Update README for MacOS/Linux

ThisJustZack committed:
Fix forgotten Note in README

ThisJustZack committed:
Refactor and add comments

Refactor to use format templates where applicable

Add comments in code to improve readability

ThisJustZack committed:
xModify docs

Modify master README to match new format

Modify plugin README to be consistent with other plugins

ThisJustZack committed:
Add hugo doc

Add hugo doc and link for plugin

ThisJustZack committed:
Correct doc deck format casing

Make the deck formats lowercase

ThisJustZack committed:
Correct hugo doc and make art file template

Fix the casing in hugo doc

Modify output files of card art to use a template

ThisJustZack committed:
Remove whitespace

ThisJustZack commented:
The one thing I have not verified yet on this plugin is if the orientation is set correctly for both sides, when needing to realign the image of a card. I'll print a double sided leader real quick to verify

ThisJustZack commented:
I just verified that it works as intended for the different rotation alignments on both sides, after doing the Chancellor Palpatine test above.

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+        back_art = None
+
+        # Ping to verify the server has the back art in one of the possible naming formats, since we do not know if it does
+        if ping_swudb(SWUDB_BACK_ART_URL_TEMPLATE_1.format(card_number=card_number_url_parse)) is True:
+            back_art = request_swudb(SWUDB_BACK_ART_URL_TEMPLATE_1.format(card_number=card_number_url_parse)).content
+        elif ping_swudb(SWUDB_BACK_ART_URL_TEMPLATE_2.format(card_number=card_number_url_parse)) is True:
+            back_art = request_swudb(SWUDB_BACK_ART_URL_TEMPLATE_2.format(card_number=card_number_url_parse)).content
```
It seems to me that this could be simplified. Right now we're pinging the database at least twice for each card.

ThisJustZack replied:
I was thinking the same, but realized that it was better to keep both, as both patterns are possible. When using the ``request`` side, it would automatically error and not make it possible to save the art if it was possible.
If the request was able to be done in the promise pattern, then I could see the ping being removed.

ThisJustZack replied:
Yeah, I agree on that. It may be better to retrieve it this way for this scenario. What could happen is using ``swu-db`` for retrieving the card name, then proceeding down the card name path via ``swudb`` for all scenarios.

ThisJustZack commented on plugins/star_wars_unlimited/README.md:
````diff
@@ -0,0 +1,279 @@
+
+## Formats
+
+### `json`
+
+The format for ``JSON``.
+
+```
+{
+  "metadata": {
+    "name": "Wichita PQ - 4th",
+    "author": "aces"
+  },
````
The JSON format is one of the three export options from swudb

Alan-Cha replied:
Ah ok. I think we should consider renaming it to `swudb_json` similar to the `scryfall_json` of the MTG plugin

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+    else: # Fetch card art by querying name and title
+        title_query = '' if title == '' else f' title:"{title}"'
+        json = request_swudb(SWUDB_NAME_URL_TEMPLATE.format(name=name, title=title_query)).json()
+        art_url_suffix = sub('.+cards/', '', json.get('variants')[0].get('frontImagePath'))
```
I think there should ideally be a check to see if there are indeed variants? The list is not empty?

ThisJustZack replied:
I was thinking the same, but I believe that this was the way it is for swudb: if the query wasn't possible then it would have a 404 error, otherwise return information

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+):
+    # Query for card art
+    if card_number != '': # Fetch card art by card number
+        card_number_url_parse = sub('_', '/', card_number)
```
I think it makes more sense to pass card number and expansion abbreviation separately, and construct the URL with these two components.

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+    else: # Fetch card art by querying name and title
+        title_query = '' if title == '' else f' title:"{title}"'
+        json = request_swudb(SWUDB_NAME_URL_TEMPLATE.format(name=name, title=title_query)).json()
+        art_url_suffix = sub('.+cards/', '', json.get('variants')[0].get('frontImagePath'))
+        front_art = request_swudb(SWUDB_ART_URL_TEMPLATE.format(card_art_ref=art_url_suffix)).content
```
I think it's strange to extract the front image path. At this point, we should be able to get the card number and expansion abbreviation. 

This is related to https://github.com/Alan-Cha/silhouette-card-maker/pull/23/files#r2280967063 because if we did that, then we can reuse the URL here.

```python
json.get('variants')[0].get('expansionAbbreviation')
json.get('variants')[0].get('cardNumber')
```

Is it also necessary to use the `get()` function here?

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+            output_file = OUTPUT_CARD_ART_FILE_TEMPLATE.format(deck_index=str(index), card_name=card_number, quantity_counter=str(counter + 1))
+
+        if front_art != None:
+            if card_number == '':
+                front_image_path = path.join(front_img_dir, output_file)
+            else:
+                front_image_path = path.join(front_img_dir, output_file)
```
What is this if statement doing?

ThisJustZack replied:
This was one of the tech debt areas I missed when refactoring for the file paths, since before the output files were determined there rather than before getting to that logic.

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+                front_image_rotated.save(front_image_path)  
+
+        if back_art != None:
+            if card_number == '':
+                back_image_path = path.join(back_img_dir, output_file)
+            else:
+                back_image_path = path.join(back_img_dir, output_file)
```
Similarly, I'm not sure what this if statement is doing

ThisJustZack replied:
This was one of the tech debt areas I missed when refactoring for the file paths, since before the output files were determined there rather than before getting to that logic.

Alan-Cha commented on plugins/star_wars_unlimited/swudb.py:
```diff
@@ -0,0 +1,117 @@
+            else:
+                front_image_path = path.join(front_img_dir, output_file)
+
+            with open(front_image_path, 'wb') as f:
+                f.write(front_art)
+
+            # Align the rotated art so that it has the correct orientation
+            front_image_for_rotation = Image.open(front_image_path)
+            if front_image_for_rotation.height < front_image_for_rotation.width:
+                front_image_rotated = front_image_for_rotation.rotate(90, expand=True)
+                front_image_rotated.save(front_image_path)  
```
Is this first writing the image file, then checking the orientation of the image, if it needs to be rotate, then it overwrites the original image file?

This seems like it could be simplified by first checking orientation, rotating the image if necessary, then saving.

ThisJustZack replied:
Yes, that was the intent, but to my knowledge Pillow requires the image to be saved before it can be utilized for rotating.

ThisJustZack committed:
Modify to use swu-db and remove tech debt

JSON uses swu-db to retrieve the card name and title rather than pinging for retrieval

All card retrievals are now by name and title

Remove tech debt of duplicate lines

ThisJustZack committed:
Modify json to be swudb_json

Alan-Cha committed:
Fix SWU plugin

Signed-off-by: Alan-Cha <Alan.cha1@ibm.com>

Alan-Cha committed:
Handle JSON separately from decklist

Signed-off-by: Alan-Cha <Alan.cha1@ibm.com>

Alan-Cha committed:
Edit Star Wars Unlimited hugo

Signed-off-by: Alan-Cha <Alan.cha1@ibm.com>

Alan-Cha committed:
Change plugin order

Signed-off-by: Alan-Cha <Alan.cha1@ibm.com>

Alan-Cha committed:
Change plugin order

Signed-off-by: Alan-Cha <Alan.cha1@ibm.com>

Alan-Cha committed:
Rename card_number to card_id

Signed-off-by: Alan-Cha <Alan.cha1@ibm.com>
